core/platform/auth/firebase_auth_services.py

```diff
@@ -0,0 +1,213 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Service layer for handling user-authentication with Firebase.
+
+Oppia depends on OpenID Connect 1.0 to handle user authentication. We use
+[Firebase authentication](https://firebase.google.com/docs/auth) to do the
+heavy-lifting, especially for securely storing user credentials and associating
+users to their identity providers. This helps us minimize the contact we make
+with private information.
+
+Terminology:
+    OpenID Connect 1.0 (OIDC):
+        A simple identity layer on top of the OAuth 2.0 protocol. It is a
+        specification (i.e. a strict set of algorithms, data structures, and
+        rules) that defines how two parties must share data about a user in
+        a secure way on that user's behalf.
+    OAuth 2.0 (OAuth):
+        The industry-standard protocol for authorization. It enables a
+        third-party application to obtain limited access to an HTTP service on
+        behalf of a user.
+    Claim:
+        A piece of information about a user (name, address, phone number, etc.)
+        that has been encrypted and digitally signed.
+    JSON Web Token (JWT):
+        A compact and URL-safe protocol primarily designed to send Claims
+        between two parties. Claims are organized into JSON objects that map
+        "Claim Names" to "Claim Values".
+    Identity provider:
+        An entity that creates, maintains, and manages identity information and
+        provides authentication services. Such services rely on JWTs to send
+        identity information. Examples of identity providers include: Google,
+        Facebook, Email verification links, and Text message SMS codes.
+    Subject Identifier:
+        A Claim that can uniquely identify a user. It is locally unique and
+        never reassigned with respect to the provider who issued it. The Claim's
+        name is 'sub'.
+        Example values: `24400320` or `AItOawmwtWwcT0k51BayewNvutrJUqsvl6qs7A4`.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import logging
+
+from core.platform import models
+import python_utils
+
+import firebase_admin
+from firebase_admin import auth as firebase_auth
+from firebase_admin import exceptions as firebase_exceptions
+
+auth_models, = models.Registry.import_models([models.NAMES.auth])
+
+
+def _ensure_firebase_is_initialized():
+    """Initializes the Firebase Admin SDK."""
+    try:
+        return firebase_admin.get_app()
+    except (ValueError, firebase_exceptions.FirebaseError):
+        pass
+
+    try:
+        return firebase_admin.initialize_app()
+    except (ValueError, firebase_exceptions.FirebaseError) as e:
+        logging.error(e)
+        raise
```

Comment:
Oh, I thought the issue was that re-raising in Python 2 with just 'raise' doesn't properly preserve the backtrace. @vojtechjelinek can you weigh in on this in case I misunderstood?

Reply:
No longer relevant since all functions return None; exceptions are not raised anymore. Marking as resolved.